dist/src/main/java/io/camunda/webapps/controllers/ForwardErrorController.java

```diff
@@ -0,0 +1,125 @@
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@Controller
+public class ForwardErrorController {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ForwardErrorController.class);
+
+  private static final String LOGIN_RESOURCE = "/api/login";
+  private static final String API = "/api/**";
+  private static final String PUBLIC_API = "/v*/**";
+  private static final String DEFAULT_MANAGEMENT_URI = "/actuator/**";
+  private static final String SSO_CALLBACK_URI = "/sso-callback";
+  private static final String NO_PERMISSION = "/noPermission";
+  private static final String IDENTITY_CALLBACK_URI = "/identity-callback";
+  private static final String REQUESTED_URL = "requestedUrl";
+
+  @Autowired private Environment environment;
+
+  private List<RequestMatcher> ignoredRequestsMatchers;
+
+  @PostConstruct
+  public void init() {
+    // list of endpoints for which 404 is returned
+    ignoredRequestsMatchers =
+        List.of(
+            new AntPathRequestMatcher(PUBLIC_API),
+            new AntPathRequestMatcher(API),
+            new AntPathRequestMatcher(NO_PERMISSION),
+            new AntPathRequestMatcher(SSO_CALLBACK_URI),
+            new AntPathRequestMatcher(IDENTITY_CALLBACK_URI),
+            // actuator endpoints
+            new AntPathRequestMatcher(
+                Optional.ofNullable(environment.getProperty("management.endpoints.web.base-path"))
+                    .map(s -> s.concat("/**"))
+                    .orElse(DEFAULT_MANAGEMENT_URI)));
+  }
+
+  @RequestMapping(value = {"/{regex:[\\w-]+}", "/**/{regex:[\\w-]+}"})
+  public String forward404(final HttpServletRequest request, final HttpServletResponse response) {
+    if (ignoredRequestsMatchers.stream()
+        .anyMatch(requestMatcher -> requestMatcher.matches(request))) {
+      response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+      return null;
+    }
+    final String requestedURL = getRequestedURL(request);
+    if (!requestedURL.startsWith("/tasklist") && !requestedURL.startsWith("/operate")) {
+      if (isStandaloneTasklist()) {
+        return "redirect:/tasklist" + requestedURL;
+      } else {
+        return "redirect:/operate" + requestedURL;
+      }
+    } else if (isLoginDelegated() && isNotLoggedIn()) {
+      return saveRequestAndRedirectToLogin(request, requestedURL);
+    } else {
+      if (isStandaloneTasklist()) {
+        return "forward:/tasklist";
+      } else {
+        return "forward:/operate";
+      }
+    }
+  }
+
+  private boolean isStandaloneTasklist() {
+    return environment.acceptsProfiles(TASKLIST.getId(), "!" + OPERATE.getId());
+  }
```

Comment:
We must get rid of such code patterns. That means we should not introduce a pattern in which we are checking the profile everywhere and, depending on it, it behaves differently. That way, we will lose the overview.

Instead, we should configure the application and make sure that the code behaves the same way.

Reply:
Done! new property `camunda.webapps.default-app` is introduced and set by `WebappsConfigurationInitializer.java`